### PR TITLE
Infered types (`_`) in `#[ts(as = "...")]`

### PR DESCRIPTION
**Pull request by escritorio-gustavo**
## Goal

This PR aims to allow the use of infered types (`_`) inside the `#[ts(as = "...")]` attribute. The `_` type will be interpreted as the field's original type. E.g.:

```rs
#[derive(TS)]
struct Foo {
    #[ts(as = "Option<_>"]
    bar: i32,
}
```

is the same as:
```rs
#[derive(TS)]
struct Foo {
    #[ts(as = "Option<i32>"]
    bar: i32,
}
```
but without having to repeat the original type.

This allows us to somewhat support using `#[ts(optional)]` with any type, by using `#[ts(as = "Option<_>", optional)]` without requiring us to change the semantics of `#[ts(optional)]`

## Changes

Added a recursive function to traverse the provided type and replace `Type::Infer(_)` with the field's original type

## Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/Aleph-Alpha/ts-rs/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Comment by NyxCode** [comment]
Interesting! So `_` refers to the original type, right?

**Comment by NyxCode** [comment]
I do think this is pretty neat, though I think it might be a bit unintuitive for anyone who doesn't know what `_` does in that context.  
Have you considered a different syntax, e.g `#[ts(as = "Option<type>")]`. I feel like we might be able to come up with something more intuitive, but I'm not sure if the extra effort/complexity involved in parsing this would be worth it.

**Comment by gustavo-shigueo** [comment]
> Interesting! So `_` refers to the original type, right?

Yes, this is mostly a way to resolve the discussion in #175

**Comment by abhay-agarwal** [comment]
I think the understore is super intuitive, as it's already widely used throughout rust syntax. So the code is instantly readable. The other nice thing is that you could, for example, turn a type into a vector, such as `as = Vec<_>`, or `as = Result<_,Something>`, etc.

**Comment by escritorio-gustavo** [comment]
There is also precedent for this behavior in `serde_with` https://github.com/jonasbb/serde_with?tab=readme-ov-file#examples

**Comment by NyxCode** [comment]
Oh, alright! Maybe it's just me then.  
I feel like the semantics of `_` are very different here, but I think you two convinced me!

File: macros/src/types/mod.rs [hunk 1]
```diff
@@ -46,3 +51,122 @@ fn type_def(attr: &StructAttr, ident: &Ident, fields: &Fields) -> Result<Derived
         Fields::Unit => unit::null(attr, &name),
     }
 }
+
+pub(super) fn type_as_infer(type_as: &Type, original_type: &Type) -> Type {
```
Comment (NyxCode) [inline macros/src/types/mod.rs:55]: This is perfectly fine, but just an idea: Would changing the signature to `type_as_infer(type_as: &Type, original: &mut Type) -> ()` make the impl cleaner? I think that might get rid of the `parse_quote!` calls.

File: macros/src/types/newtype.rs [hunk 2]
```diff
@@ -27,7 +28,10 @@ pub(crate) fn newtype(attr: &StructAttr, name: &str, fields: &FieldsUnnamed) ->
         return super::unit::null(attr, name);
     }
 
-    let inner_ty = type_as.as_ref().unwrap_or(&inner.ty).clone();
+    let inner_ty = type_as
+        .as_ref()
+        .map(|ty_as| type_as_infer(ty_as, &inner.ty))
+        .unwrap_or_else(|| inner.ty.clone());
```
Comment (NyxCode) [inline macros/src/types/newtype.rs:34]: Would it be possible to move this into a getter on `FieldAttr`, making the field `type_as` private?

Comment (escritorio-gustavo) [reply macros/src/types/newtype.rs:34]: There's also the `format_variant` function that needs access to `type_as`

File: macros/src/types/mod.rs (hunk 1 above)
Comment (NyxCode) [reply macros/src/types/mod.rs:55]: I've put something together in #305 to highlight what I had in mind. I've put the `&mut` on the wrong argument in the snippet above, i think that caused some confusion. Sorry about that.

Comment (escritorio-gustavo) [reply macros/src/types/mod.rs:55]: > I've put something together in https://github.com/Aleph-Alpha/ts-rs/pull/305 to highlight what I had in mind.

Oh, that makes a lot more sense! I'll merge that in real quick

File: macros/src/types/enum.rs [hunk 3]
```diff
@@ -135,30 +135,16 @@ fn format_variant(
...
-                        (Some(type_override), None) => quote! { #type_override },
-                        (None, Some(type_as)) => {
-                            quote!(<#type_as as #crate_rename::TS>::name())
-                        }
-                        (None, None) => {
-                            let ty = &unnamed.unnamed[0].ty;
+                    let ty = match field_attr.type_override {
+                        Some(type_override) => quote!(#type_override),
+                        None => {
+                            let ty = field_attr.type_as(&field.ty);
```
Comment (NyxCode) [inline macros/src/types/enum.rs:144]: I love how this got a lot simpler. Awesome.

**Review by NyxCode — APPROVED** [review]
Awesome stuff, great work!

**Comment by NyxCode** [comment]
Also, you were spot on with your take in #175. This solution keeps `#[ts(optional)]` as it was. Specifically, `#[ts(optional)]` still doesn't alter the type, just how it's represented. This seems super clean to me!

**Comment by NyxCode** [comment]
Interestingly, this kinda makes `#[ts(optional = nullable)]` obsolete. It'd be equivalent to 
```rs
#[ts(as = "Option<_>", optional)]
field: Option<i32>,
```

**Pull request merged**